File: elixir/list-ops/lib/list_ops.ex
defmodule ListOps do
  def count(list \\ [], i \\ 0)
  def count([],i), do: i
  def count([_h | t], i), do: count(t, i+1)
  def reverse(list), do: list |> reduce([], &[&1| &2])
  def map(list, fun), do: list |> reverse() |> reduce([], &[fun.(&1) | &2])
  @spec filter(list, (any -> as_boolean(term))) :: list
  def filter(list \\ [], f \\ fn x -> x == 0 end, newlist \\ [])
  def filter([], _f, newlist), do: reverse(newlist)
  def filter([h | t ], f, newlist) do
    if f.(h) do
      filter(t, f, [h | newlist])
    else
      filter(t, f, newlist)
    end
  end
  def reduce([], acc, _fun), do: acc
  def reduce([h | t], acc, fun), do: h |> fun.(acc) |> (&(reduce(t, &1, fun))).()
  def append(a, b), do: a |> reverse |> reduce(b, &[&1 | &2])
  def concat(a), do: a |> reverse |> reduce([], &append(&1, &2))
end

File: javascript/list-ops/list-ops.js
export class List {
  constructor(list = []) {
    this.values = list;
  }

  length() {
    for (var i = 0; this.values[i]; i++) {}
    return i;
  }

  append(list) {
    return new List([...this.values, ...list.values]);
  }

  concat(lists) {
    return this.append(this._iterateObj(lists));
  }

  reverse() {
    return new List(this._reduce([], (reversed, next) => [next, ...reversed]));
  }

  map(fun) {
    for (let i = 0; this.values[i]; ++i) {
      this.values[i] = fun(this.values[i]);
    }
    return this;
  }

  filter(fun) {
    let filter = [];
    let filterList = new List([]);
    for (let i = 0; this.values[i]; ++i) {
      filter[i] = Number(this.values.map(fun)[i]);
      if (filter[i]) filterList = filterList._appendOne(this.values[i]);
    }
    return filterList;
  }

  foldl(fun, acc) {
    return this._reduce(acc, fun);
  }

  foldr(fun, acc) {
    return this.reverse()._reduce(acc, fun);
  }

  // Helper functions
  _appendOne(item) {
    return new List([...this.values, item]);
  }

  _iterateObj(obj, list = new List()) {
    for (let i = 0; obj.length() > i; ++i) {
      list = list.append(obj.values[i]);
    }
    return list;
  }

  _reduce(acc, fun) {
    for (let next of this.values) {
      acc = fun(acc, next);
    }
    return acc;
  }
}
